Library/Homebrew/cask/cask_loader.rb

```diff
@@ -186,6 +186,100 @@ def load(config:)
       end
     end
 
+    # Loads a cask from the JSON API.
+    class FromAPILoader
+      attr_reader :token, :path
+
+      FLIGHT_STANZAS = [:preflight, :postflight, :uninstall_preflight, :uninstall_postflight].freeze
+
+      def self.can_load?(ref)
+        Homebrew::API::Cask.all_casks.key? ref
+      end
+
+      def initialize(token)
+        @token = token
+        @path = CaskLoader.default_path(token)
+      end
+
+      def load(config:)
+        json_cask = Homebrew::API::Cask.all_casks[token]
+
+        if (bottle_tag = ::Utils::Bottles.tag.to_s.presence) &&
+           (variations = json_cask["variations"].presence) &&
+           (variation = variations[bottle_tag].presence)
+          json_cask = json_cask.merge(variation)
+        end
+
+        json_cask.deep_symbolize_keys!
+
+        # Use the cask-source API if there are any `*flight` blocks
+        if json_cask[:artifacts].any? { |artifact| FLIGHT_STANZAS.include?(artifact.keys.first) }
+          cask_source = Homebrew::API::CaskSource.fetch(token)
```

Comment:
```suggestion
          cask_source = Homebrew::API::CaskSource.fetch(token, sha)
```

Not for this PR: I think it would be good for this API to require fetching the cask source at a particular Git SHA/revision that matches that from the JSON, for consistency and to be able to use the e.g. https://raw.githubusercontent.com/Homebrew/homebrew-cask/25e4143c41a98abb50cde6a5042e957fe5988236/Casks/0-ad.rb endpoints.

Reply:
@Rylan12 Yeh, I think so. Some advantages:
- don't need to duplicate these files
- can ensure that we store e.g. the homebrew/cask revision in the "all casks" `.json` file and that we are always querying the correct/consistent version of the cask based on this revision